firmware/application/apps/ui_text_editor.cpp

```diff
@@ -221,28 +222,45 @@ void TextViewer::paint_text(Painter& painter, uint32_t line, uint16_t col) {
                  r.top() + (int)i * char_height,
                  clear_width * char_width, char_height},
                 style().background);
+
+        // if cursor line got overwritten, disable XOR of old cursor when displaying new cursor
+        if (i == paint_state_.line)
+            paint_state_.line = UINT32_MAX;
     }
 }
 
 void TextViewer::paint_cursor(Painter& painter) {
     if (!has_focus())
         return;
 
-    auto draw_cursor = [this, &painter](uint32_t line, uint16_t col, Color c) {
+    auto xor_cursor = [this, &painter](int32_t line, uint16_t col) {
         auto r = screen_rect();
-        line = line - paint_state_.first_line;
-        col = col - paint_state_.first_col;
-
-        painter.draw_rectangle(
-            {(int)col * char_width - 1,
-             r.top() + (int)line * char_height,
-             char_width + 1, char_height},
-            c);
+        int cursor_width = char_width + 1;
+
+        int x = (col - paint_state_.first_col) * char_width - 1;
+        if (x < 0) {  // cursor is one pixel narrower when in left column
+            cursor_width--;
+            x = 0;
+        }
+
+        int y = r.top() + (line - paint_state_.first_line) * char_height;
+
+        int cursor_pixels = cursor_width * char_height;
+        ColorRGB888 pbuf888[cursor_pixels];
+        Color pbuf[cursor_pixels];
+
+        // annoyingly, read_pixels uses a 24-bit pixel format vs draw_pixels which uses 16-bit
```

Comment:
Can we fix that? That seems lame. Doesn't have to be for this PR, though.

Reply:
Deferring to future PR.  It's probably simple but requires study of the ILI9341 specs since there's currently no function written to read a pixel in the 16-bit format.